Sources/StreamChat/Utils/Database/NSManagedObject+Extensions.swift

```diff
@@ -163,3 +163,42 @@ extension NSManagedObjectContext {
         return objects
     }
 }
+
+// MARK: - Runtime Filtering
+
+extension Set where Element: NSManagedObject {
+    func filtered(using request: NSFetchRequest<Element>) -> [Element] {
+        guard !isEmpty else { return [] }
+        var allObjects = NSMutableArray(array: (self as NSSet).allObjects, copyItems: false)
+        if let sortDescriptors = request.sortDescriptors, !sortDescriptors.isEmpty {
+            allObjects.sort(using: sortDescriptors)
+        }
+        if request.fetchLimit > 0 {
+            let result = NSMutableArray(capacity: request.fetchLimit)
+            let batchSize = Swift.min(request.fetchLimit, 100)
+            for index in stride(from: 0, to: allObjects.count, by: batchSize) {
+                let range = NSRange(location: index, length: Swift.min(batchSize, allObjects.count - index))
+                let batch = allObjects.subarray(with: range)
+                if let predicate = request.predicate {
+                    result.addObjects(from: (batch as NSArray).filtered(using: predicate))
+                } else {
+                    result.addObjects(from: batch)
+                }
+                if result.count > request.fetchLimit {
+                    let removeCount = result.count - request.fetchLimit
+                    result.removeObjects(in: NSRange(location: result.count - removeCount, length: removeCount))
+                }
+                if result.count == request.fetchLimit {
+                    break
+                }
+            }
+            allObjects = result
+        } else {
+            if let predicate = request.predicate {
+                allObjects.filter(using: predicate)
+            }
+        }
+        
+        return allObjects as? [Element] ?? []
+    }
```

Comment:
I would really like to remove this, all the fetch request logic should only be used on the `FetchResultsController`, the Immutable Models, should just get data from the backend, and not have predicates

Reply:
If we go for removing the messages fetch, then I am happy to delete this. Will look into this tomorrow, but I feel like it makes sense.